zebra-chain/src/block.rs

```diff
@@ -104,22 +115,34 @@ pub struct BlockHeader {
     nonce: [u8; 32],
 
     /// The Equihash solution.
-    // The solution size when serialized should be in bytes ('always
-    // 1344').  I first tried this as a [u8; 1344] but until const
-    // generics land we'd have to implement all our common traits
-    // manually, like in pzec.
-    solution: Vec<u8>,
+    solution: EquihashSolution,
 }
 
 impl ZcashSerialize for BlockHeader {
-    fn zcash_serialize<W: io::Write>(&self, _writer: W) -> Result<(), SerializationError> {
-        unimplemented!();
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+        self.previous_block_hash.zcash_serialize(&mut writer)?;
+        writer.write_all(&self.merkle_root_hash.0[..])?;
+        writer.write_all(&self.final_sapling_root_hash.0[..])?;
```

Comment:
Per https://zips.z.cash/protocol/protocol.pdf#blockheader 

> [Pre-Sapling] A reserved field which should be ignored. [Sapling onward] The root LEBS2OSP256(rt) of the Sapling note commitment tree corresponding to the final Sapling treestate of this block.

Is it worth doing something to represent this internally, for instance having pre-Sapling and post-Sapling enum variants? I don't feel great about not having all of the information required for semantic validity in the block itself. On the other hand, they are actually the same version, and it may just be "like that".

Reply:
(Note: this is an observation I noticed while trying to check that the serialize implementation matches the spec, but it actually has nothing to do with the serialization itself. So it could be resolved separately.)

Reply:
> Is it worth doing something to represent this internally, for instance having pre-Sapling and post-Sapling enum variants?

I would feel better about doing this too.

Reply:
I feel a bit awkward making an internal enum variant that doesn't correspond to a version -- perhaps it's not a problem to have every block keep a final sapling treestate and just ignore it when there is no sapling data present?